src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1389 @@
+                //         using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutVoidAsyncMethod, "var x = new C()").WithLocation(6, 22)
+                );
+        }
+
+        [Fact]
+        public void TestInNonAsyncMethodReturningInt()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    int M()
+    {
+        using await (var x = new C())
+        {
+            return 1;
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (6,22): error CS4032: The 'await' operator can only be used within an async method. Consider marking this method with the 'async' modifier and changing its return type to 'Task<int>'.
+                //         using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutAsyncMethod, "var x = new C()").WithArguments("int").WithLocation(6, 22)
+                );
+        }
+
+        [Fact]
+        public void TestInNonAsyncAnonymousMethod()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    void M()
+    {
+        System.Action x = () =>
+        {
+            using await (var y = new C())
+            {
+                return;
+            }
+        };
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (8,26): error CS4034: The 'await' operator can only be used within an async lambda expression. Consider marking this lambda expression with the 'async' modifier.
+                //             using await (var y = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutAsyncLambda, "var y = new C()").WithArguments("lambda expression").WithLocation(8, 26)
+                );
+        }
+
+        [Fact]
+        public void TestInAsyncAnonymousMethod()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    C()
+    {
+        System.Console.Write(""C "");
+    }
+    public static void Main()
+    {
+        System.Action x = async () =>
+        {
+            using await (var y = new C())
+            {
+                System.Console.Write(""body "");
+                return;
+            }
+        };
+        x();
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        System.Console.Write(""DisposeAsync"");
+        return System.Threading.Tasks.Task.CompletedTask;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "C body DisposeAsync");
```

Comment:
>"C body DisposeAsync" [](start = 51, length = 21)

It would be good to test that the task is actually awaited, which I believe this test doesn't confirm.